zipkin-lens/src/components/TracePage/TraceSummary.jsx

```diff
@@ -11,72 +11,183 @@
 };
 
 const TraceSummary = ({ traceSummary }) => {
-  const [tsRange, setTsRange] = useState({
-    startTs: 0,
-    endTs: traceSummary.duration,
-  });
-
-  const handleStartAndEndTsChange = useCallback((startTs, endTs) => {
-    setTsRange({
-      startTs,
-      endTs,
+  const isRootedTrace = hasRootSpan(traceSummary.spans);
+  const [rootSpanIndex, setRootSpanIndex] = useState(0);
+  const isRerooted = rootSpanIndex !== 0;
+  const [currentSpanIndex, setCurrentSpanIndex] = useState(0);
+  const [childrenHiddenSpanIds, setChildrenHiddenSpanIds] = useState({});
+  const [isSpanDetailOpened, setIsSpanDetailOpened] = useState(true);
+  const traceTimelineWidthPercent = isSpanDetailOpened ? 60 : 100;
+
+  const handleChildrenToggle = useCallback((spanId) => {
+    setChildrenHiddenSpanIds(prevChildrenHiddenSpanIds => ({
+      ...prevChildrenHiddenSpanIds,
+      [spanId]: prevChildrenHiddenSpanIds[spanId] ? undefined : true,
+    }));
+  }, []);
+
+  const handleResetRerootButtonClick = useCallback(() => {
+    setRootSpanIndex(0);
+  }, []);
+
+  const handleTimelineRowClick = useCallback((spanId) => {
+    const idx = traceSummary.spans.findIndex(span => span.spanId === spanId);
+    if (isRootedTrace && currentSpanIndex === idx) {
+      if (rootSpanIndex === idx) {
+        setRootSpanIndex(0);
+      } else {
+        setRootSpanIndex(idx);
+      }
+    }
+    setCurrentSpanIndex(idx);
+    setIsSpanDetailOpened(true);
+  }, [currentSpanIndex, isRootedTrace, traceSummary.spans, rootSpanIndex]);
+
+  const rerootedTree = useMemo(() => {
+    // If the trace does not have a root span, the trace is not filtered anymore
+    // and the entire trace should be displayed.
+    if (!isRootedTrace) {
+      return traceSummary.spans;
+    }
+
+    const rootSpan = traceSummary.spans[rootSpanIndex];
+    const spans = [rootSpan];
+    for (let i = rootSpanIndex + 1; i < traceSummary.spans.length; i += 1) {
+      const span = traceSummary.spans[i];
+      if (span.depth <= rootSpan.depth) {
+        break;
+      }
+      spans.push(span);
+    }
+    return spans;
+  }, [isRootedTrace, rootSpanIndex, traceSummary.spans]);
+
+  const shownTree = useMemo(() => {
+    const allHiddenSpanIds = {};
+    rerootedTree.forEach((span) => {
+      if (childrenHiddenSpanIds[span.parentId]) {
+        allHiddenSpanIds[span.spanId] = true;
+      }
+      if (allHiddenSpanIds[span.spanId] && span.childIds) {
+        span.childIds.forEach((childId) => {
+          allHiddenSpanIds[childId] = true;
+        });
+      }
+    });
+    return rerootedTree.filter(span => !allHiddenSpanIds[span.spanId]);
+  }, [rerootedTree, childrenHiddenSpanIds]);
+
+  // Find the minumum and maximum timestamps in the shown spans.
+  const startTs = useMemo(() => minBy(rerootedTree, 'timestamp').timestamp, [rerootedTree]);
+  const endTs = useMemo(() => {
+    let max = 0;
+    rerootedTree.forEach((span) => {
+      let ts;
+      if (!span.duration) {
```

Comment:
Just me, but I find this a bit more readable / less complexity as

```ts
let ts = span.timestamp;
if (span.duration) {
  ts += span.duration;
}
```

Reply:
Though incidentally, is it possible for `duration` to be undefined (as opposed to always being returned as 0 by API)?

Reply:
in our model we have duration only valid with minimum value 1 for reasons like this :)

Reply:
At least I found spans that don't have a duration in teatdata.

https://github.com/openzipkin/zipkin/blob/2c353e17c0a2fa8440fe01ba924aff36dd416aa3/zipkin-lens/testdata/netflix.json#L2-L41